src/main/java/org/checkerframework/checker/crypto/CryptoAnnotatedTypeFactory.java

```diff
@@ -74,12 +78,73 @@ public boolean isSubtype(final AnnotationMirror subtype, final AnnotationMirror
+            if (AnnotationUtils.areSameByName(a2, UNKNOWNALGORITHMORPROVIDER)) {
+                return a1;
+            }
+            if ((AnnotationUtils.areSameByName(a1, ALLOWEDALGORITHMS)
+                            && AnnotationUtils.areSameByName(a2, ALLOWPROVIDERS))
+                    || (AnnotationUtils.areSameByName(a2, ALLOWEDALGORITHMS)
+                            && AnnotationUtils.areSameByName(a1, ALLOWPROVIDERS))) {
+                return CryptoAnnotatedTypeFactory.this.BOTTOM;
+            }
+
+            // Both a1 and a2 are the same annotation kind.
+            ExecutableElement element =
+                    TreeUtils.getMethod(
+                            AnnotationUtils.annotationMirrorToClass(a1), "value", 0, processingEnv);
+            List<String> a1RegexList =
+                    AnnotationUtils.getElementValueArray(a1, element, String.class);
+            List<String> a2RegexList =
+                    AnnotationUtils.getElementValueArray(a2, element, String.class);
+            // Cal Intersection of a1RegexList and a2RegexList
+            a1RegexList.retainAll(a2RegexList);
+            return createCryptoAnnotation(
+                    a1RegexList,
+                    (Class<? extends Annotation>) AnnotationUtils.annotationMirrorToClass(a1));
         }
 
         @Override
         public AnnotationMirror leastUpperBound(AnnotationMirror a1, AnnotationMirror a2) {
-            return CryptoAnnotatedTypeFactory.this.UNKNOWNALGORITHMORPROVIDER;
+            if (AnnotationUtils.areSameByName(a1, UNKNOWNALGORITHMORPROVIDER)
+                    || AnnotationUtils.areSameByName(a2, UNKNOWNALGORITHMORPROVIDER)) {
+                return CryptoAnnotatedTypeFactory.this.UNKNOWNALGORITHMORPROVIDER;
+            }
+            if (AnnotationUtils.areSameByName(a1, BOTTOM)) {
+                return a2;
+            }
+            if (AnnotationUtils.areSameByName(a2, BOTTOM)) {
+                return a1;
+            }
+            if ((AnnotationUtils.areSameByName(a1, ALLOWEDALGORITHMS)
+                            && AnnotationUtils.areSameByName(a2, ALLOWPROVIDERS))
+                    || (AnnotationUtils.areSameByName(a2, ALLOWEDALGORITHMS)
+                            && AnnotationUtils.areSameByName(a1, ALLOWPROVIDERS))) {
+                return CryptoAnnotatedTypeFactory.this.UNKNOWNALGORITHMORPROVIDER;
+            }
+
+            // Both a1 and a2 are the same annotation kind.
+            ExecutableElement element =
+                    TreeUtils.getMethod(
+                            AnnotationUtils.annotationMirrorToClass(a1), "value", 0, processingEnv);
+            List<String> a1RegexList =
+                    AnnotationUtils.getElementValueArray(a1, element, String.class);
+            List<String> a2RegexList =
+                    AnnotationUtils.getElementValueArray(a2, element, String.class);
+            // Merge a1RegexList and a2RegexList, then remove duplicate elements
```

Comment:
This comment is good, it helps realize why you're having the extra Set.

Reply:
Thanks!